app/controllers/signatures_controller.rb

```diff
@@ -121,5 +111,25 @@ def handle_new_signature(petition)
       render :new
     end
   end
+
+  def validate_sponsor
+    if @signature.validated?
+      flash[:notice] = "Thank you. You have already sponsored this petition."
+    else
+      @signature.validate!
+      send_sponsor_support_notification_email_to_petition_owner(@signature.petition, @signature)
+      @signature.petition.update_state_after_new_validated_sponsor!
+    end
+    redirect_to sponsored_petition_sponsor_url(@signature.petition, token: @signature.petition.sponsor_token)
+  end
+
+  def validate_signature
+    if @signature.validated?
+      flash[:notice] = "Thank you. Your signature has already been added to the petition."
+    else
+      @signature.validate!
+    end
+    redirect_to signed_petition_signature_url(@signature.petition, @signature.perishable_token)
+  end
```

Comment:
This is nice; getting rid of the ``and return`` statements that terminate control flow in the middle of the methods makes me happy :)